awscli/autoprompt/core.py

```diff
@@ -29,47 +32,54 @@ class AutoPromptDriver:
 
     NO_PROMPT_ARGS = ['help', '--version']
 
-    def __init__(self, session, prompter=None):
+    def __init__(self, driver, completion_source=None, prompter=None):
+        self._completion_source = completion_source
         self._prompter = prompter
-        self._session = session
-
-    def _should_autoprompt(self, parsed_args, args):
+        self._session = driver.session
+        self._driver = driver
+        if self._completion_source is None:
+            self._completion_source = create_autocompleter(driver=self._driver)
+
+    @property
+    def prompter(self):
+        if self._prompter is None:
+            self._prompter = AutoPrompter(self._completion_source,
+                                          self._driver)
+        return self._prompter
+
+    def _filter_out_options(self, args):
+        override_options = ['--cli-auto-prompt', '--no-cli-auto-prompt']
+        return [
+                arg for arg in args
+                if (not arg.startswith('--') or arg in override_options)
+            ]
+
+    def resolve_mode(self, args):
         # Order of precedence to check:
         # - check if any arg rom NO_PROMPT_ARGS in args
         # - check if '--no-cli-auto-prompt' was specified
         # - check if '--cli-auto-prompt' was specified
         # - check configuration chain
+        option_parser = ArgTableArgParser(self._driver.arg_table)
+        filtered_args = self._filter_out_options(args)
+        parsed_options, _ = option_parser.parse_known_args(filtered_args)
+        validate_auto_prompt_args_are_mutually_exclusive(parsed_options)
```

Comment:
So one thing that I'm wonder if we should just roll out our own basic argument parsing for `--cli-auto-prompt` and `--no-cli-auto-prompt`? Specifically we would just validate both arguments are not in the args provided and just return `on` or `off` based on the argument provided. Looking over the driver code: https://github.com/aws/aws-cli/blob/ab41dc3429137dd964670e9bc342590ca4b7bf2b/awscli/clidriver.py#L319-L331 I'm a little concerned that we would emitting `building-top-level-params` out of order of the typical order it gets invoked and I'm not sure if there will be unknown side effects. Like I think this event might be typically emitted before it: https://github.com/aws/aws-cli/blob/ab41dc3429137dd964670e9bc342590ca4b7bf2b/awscli/clidriver.py#L298

Reply:
On a similar note, it may make sense to update the functional tests to start using the new `AWSCLIEntryPoint` to see if points out any issues related to my concern.

Reply:
Hmmmm. good point, no idea why we did it this way